### DISCUSSION TITLE
Embed micropython in an Arduino project

### DISCUSSION THREAD

**Discussion by jhdyz** [micropython/micropython#11172] [General]
I want to embed micropython in an Arduino project.

I see the many works about this topic have been done in micropython/examples/embedding and micropython/ports/embed . 

The example successfully embeds micropython in a standalone C application , but i stiil don't know how to embed it in an arduino project to esp32

Can any one give me any suggestions or someone has try this before?

really thanks

**Comment by Josverl** [comment]
I do not think that makes sense, but perhaps I do not understand the scenario or problem you are out to solve.
Also the usual of MCUs in the typical Arduino project does not have the capability's needed to run MicroPython. 

If you explain what type of 'Arduino project' you have in mind, it may help to understand what solution could fit.

**Reply by jhdyz** [reply]
thanks, i am sorry and here is more details:
I use VScode+platformIO to create a project, it uses Arduino framework and targets to Espressif esp32 dev borad
the tree of the project is:
```py
.pio
.vscode
.include
.lib
.src
----main.cpp
.test
.gitignore
platformio.ini
```
in the main.cpp i copy the example code, and i want to successfully build this project and upload it to esp32
```cpp
#include <Arduino.h>
#include "port/micropython_embed.h"

static const char *example_1 =
    "print('hello world!', list(x + 1 for x in range(10)), end='eol\\n')";

static const char *example_2 =
    "for i in range(10):\n"
    "    print('iter {:08}'.format(i))\n"
    "\n"
    "try:\n"
    "    1//0\n"
    "except Exception as er:\n"
    "    print('caught exception', repr(er))\n"
    "\n"
    "print('finish')\n"
    ;

static char heap[8 * 1024];

void setup() {
    mp_embed_init(&heap[0], sizeof(heap));

    mp_embed_exec_str(example_1);
    mp_embed_exec_str(example_2);

    mp_embed_deinit();
}

void loop() {}
```
I try to include the micropython files that's needed, when i build the project, the error infomation is
```cpp
Linking .pio\build\esp32dev\firmware.elf
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: .pio\build\esp32dev\src\main.cpp.o:(.literal._Z5setupv+0x10): undefined reference to `mp_embed_init(void*, unsigned int)'
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: .pio\build\esp32dev\src\main.cpp.o:(.literal._Z5setupv+0x14): undefined reference to `mp_embed_exec_str(char const*)'
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: .pio\build\esp32dev\src\main.cpp.o:(.literal._Z5setupv+0x18): undefined reference to `mp_embed_deinit()'
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: .pio\build\esp32dev\src\main.cpp.o: in function `setup()':
E:\Code\MicroPythonEmbedded2/src/main.cpp:26: undefined reference to `mp_embed_init(void*, unsigned int)'
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: E:\Code\MicroPythonEmbedded2/src/main.cpp:29: undefined reference to `mp_embed_exec_str(char const*)'
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: E:\Code\MicroPythonEmbedded2/src/main.cpp:30: undefined reference to `mp_embed_exec_str(char const*)'
c:/users/administrator/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: E:\Code\MicroPythonEmbedded2/src/main.cpp:33: undefined reference to `mp_embed_deinit()'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\esp32dev\firmware.elf] Error 1
```
I have included micropython_embed.h and i don't know why it throws "undefined reference to xxx_function"

**Comment by jhdyz** [comment]
I find the reason that  throws "undefined reference to xxx_function" ,because 
```cpp
as an object-oriented language, C + + supports function overloading, while  C does not. The names of functions compiled in C + + in the symbol library are different from those in C. 
```
I use main.cpp and include C files , if i want to make it correct,  i should use extern "C" ,here is the main.cpp (I delete the part of gc collect):
```cpp
#include <Arduino.h>
extern "C"{
    #include "port/micropython_embed.h"
}

static const char *example_1 =
    "print('hello world!', list(x + 1 for x in range(10)), end='eol\\n')";

static const char *example_2 =
    "for i in range(10):\n"
    "    print('iter {:08}'.format(i))\n"
    "\n"
    "try:\n"
    "    1//0\n"
    "except Exception as er:\n"
    "    print('caught exception', repr(er))\n"
    "\n"
    "print('finish')\n"
    ;


void setup() {
    mp_embed_init();

    mp_embed_exec_str(example_1);
    mp_embed_exec_str(example_2);

    mp_embed_deinit();
}

void loop() {}

```
after i solve this problem,i met others problems but i solve them, so now I successfully embeded micropython in an Arduino project.
thanks .

**Reply by gabrielerandazzo** [reply]
Hi @jhdyz, I tried following the steps you outlined to solve the problem, but unfortunately, I continue to encounter the error: 
`micropython_embed/micropython_embed.ino:25: undefined reference to 'mp_embed_init'`
 Could you please provide more detailed guidance on your approach and any other issues you encountered?  Thank you.

**Reply by Legend-gamer983** [reply]
Same here, I also want to know the solution to the problem as @gabrielerandazzo mentioned.
If you could provide how did you include the C files for the project, it would mean a lot, Thank you.
